### PR TITLE
Record search history

### PR DESCRIPTION
**Pull request by MaryJml**
Fixes https://github.com/JabRef/jabref/issues/7906
截屏2023-03-12 18 52 24

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/gui/StateManager.java [hunk 1]
```diff
@@ -201,4 +206,30 @@ public LastAutomaticFieldEditorEdit getLastAutomaticFieldEditorEdit() {
...
+    public void addSearchHistory(String search) {
+        searchHistory.add(search);
+    }
+
+    public ObservableSet<String> getWholeSearchHistory() {
+        return searchHistory;
+    }
+
+    public List<String> getLastSearchHistory(int size) {
+        List<String> lastSearches = new ArrayList<>();
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/StateManager.java:228]: Could you check if there can some code be shared as a new list type with the FileHistory class? I remember that I worked some time ago with a similar mechanic to limit the max number of entries in the list.

File: src/main/java/org/jabref/gui/search/SearchFieldRightClickMenu.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    private static Menu createSearchFromHistorySubMenu(ActionFactory factory,
+                                                       StateManager stateManager,
+                                                       CustomTextField searchField) {
+        Menu searchFromHistorySubMenu = factory.createMenu(() -> "Search from history...");
+
+        int num = stateManager.getLastSearchHistory(10).size();
+        if (num == 0) {
+            MenuItem item = new MenuItem("your search history is empty");
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/search/SearchFieldRightClickMenu.java:48]: See above, also please with localization

File: src/main/java/org/jabref/gui/StateManager.java [hunk 3]
```diff
@@ -71,6 +71,8 @@ public class StateManager {
 
     private final ObjectProperty<LastAutomaticFieldEditorEdit> lastAutomaticFieldEditorEdit = new SimpleObjectProperty<>();
 
+    private final ObservableList<String> searchHistory = FXCollections.observableList(new ArrayList<>());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/StateManager.java:74]:
```suggestion
    private final ObservableList<String> searchHistory = FXCollections.obserableArrayList());
```

File: src/main/java/org/jabref/gui/StateManager.java [hunk 4]
```diff
@@ -212,4 +214,31 @@ public List<String> collectAllDatabasePaths() {
...
+        searchHistory.add(search);
+    }
+
+    public ObservableList<String> getWholeSearchHistory() {
+        return searchHistory;
+    }
+
+    public List<String> getLastSearchHistory(int size) {
+        List<String> lastSearches = new ArrayList<>();
+        Object[] searchArray = searchHistory.toArray();
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/StateManager.java:229]: You no longer need the array when you have a List

**Review by Siedlerchr — APPROVED** [review]

**Comment by Siedlerchr** [comment]
Tested it locally works fine!

**Review by Siedlerchr — APPROVED** [review]

**Comment by Siedlerchr** [comment]
I would suggest we merge this current as is, and I will move @HoussemNasri  comment to a follow up issue

**Pull request merged**